File: absolute-zero/examples/ruby/nop.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Ruby CNO (Code with No Output) - Absolute Zero Example
#
# This program executes successfully without producing any output.
# Demonstrates the minimal Ruby program that does nothing observably.
#
# INTERPRETER BEHAVIOR:
# - MRI (Matz's Ruby Interpreter) performs environment setup
# - YARV bytecode compilation (Ruby 1.9+)
# - Standard library autoload triggers are registered
# - Runtime overhead: ~20-50ms for interpreter startup
#
# RUNTIME OVERHEAD:
# - Object space initialization
# - Core classes loaded (Object, Class, Module, etc.)
# - $LOAD_PATH constructed from gem paths
# - Encoding tables initialized (UTF-8 default in Ruby 2.0+)
# - Approximate cost: 10-20 MB memory footprint minimum
#
# SIDE EFFECTS:
# - Global variables populated: $PROGRAM_NAME, $0, ARGV, etc.
# - File descriptors opened: STDIN, STDOUT, STDERR
# - Process exit code 0 set implicitly
# - Signal handlers initialized
# - ObjectSpace contains all core objects
# - No .rbc files created (unlike Python's .pyc)
#
# VERIFICATION:
# - No stdout/stderr output
# - Exit code 0
# - No exceptions raised
# - No files modified
#
# EXECUTION:
#     ruby nop.rb
#     echo $?  # Should output 0
#
# LANGUAGE SEMANTICS:
# - Empty file is valid Ruby program
# - nil is implicit return value
# - Comments-only file is valid
# - No explicit 'exit 0' needed

# Explicit no-operation (uncomment to demonstrate):
# nil

# Alternative explicit no-ops in Ruby:
# -> {}
# proc {}
# begin; end
# if false; end

File: absolute-zero/examples/perl/nop.pl
#!/usr/bin/env perl
use strict;
use warnings;

# Perl CNO (Code with No Output) - Absolute Zero Example
#
# This program executes successfully without producing any output.
# Demonstrates the minimal Perl program that does nothing observably.
#
# INTERPRETER BEHAVIOR:
# - Perl interpreter (typically perl5) initialization
# - Script compilation to internal opcodes
# - @INC constructed from system and environment paths
# - Symbol table setup for main:: package
# - Runtime overhead: ~5-20ms for Perl startup
#
# RUNTIME OVERHEAD:
# - Core Perl runtime loaded (~2-5 MB baseline)
# - Built-in variables initialized: $0, @ARGV, %ENV, etc.
# - Special variables: $/, $\, $", $;, etc. all set to defaults
# - UNIVERSAL package methods available
# - DynaLoader for loading XS modules initialized
# - Approximate cost: 5-10 MB memory footprint minimum
#
# SIDE EFFECTS:
# - Exit code 0 set implicitly
# - STDIN, STDOUT, STDERR filehandles opened
# - Current working directory cached ($ENV{PWD})
# - @INC populated with library search paths
# - %INC tracks loaded modules (empty if no use/require)
# - Signal handlers set to defaults
# - $^T (program start time) set
# - $$ (process ID) available
#
# VERIFICATION:
# - No stdout/stderr output
# - Exit code 0
# - No warnings or errors (with use strict; use warnings;)
# - No files modified
# - $? is 0 (child process status)
#
# EXECUTION:
#     perl nop.pl
#     echo $?  # Should output 0
#
# LANGUAGE SEMANTICS:
# - Empty program block after pragmas is valid
# - undef is implicit return value
# - 'use strict' and 'use warnings' compile-time directives
# - No main() function needed (unlike C)
# - exit(0) is implicit; explicit exit unnecessary
#
# PERL-SPECIFIC NOTES:
# - BEGIN blocks would execute at compile time
# - END blocks would execute at exit
# - CHECK/INIT blocks for special phase execution
# - Taint mode (-T) would be enforced if enabled
# - No .plc compiled cache (unlike Python/Ruby bytecode)

# Explicit no-operation (uncomment to demonstrate):
# 1;

# Alternative explicit no-ops in Perl:
# ();
# {};
# undef;
# sub nop {}
# package Nop;
# BEGIN {}
# END {}

# Note: Perl modules conventionally end with '1;' to indicate
# successful loading, but scripts do not require this.

__END__

# Everything after __END__ is ignored by Perl
# This is a valid way to add documentation or data sections
# that won't be executed.
